client/src/App.tsx

```diff
@@ -361,37 +361,40 @@ function App() {
 		}
 	}, []);
 
-	const fireTowers = useCallback(() => {
-		towerDefenseState.towers.forEach((tower) => {
-			// only hit first enemy
-			for (let i = 0; i < towerDefenseState.units.length; i++) {
-				const unit = towerDefenseState.units[i];
-
-				const { ref } = unit;
-				if (ref && ref.current) {
-					const rect = ref.current.getBoundingClientRect();
-
-					const distance = getDistanceBetweenPoints(
-						tower.left,
-						tower.top,
-						rect.left,
-						rect.top
-					);
-
-					const relativeDistance = distance / window.innerWidth;
-
-					if (relativeDistance < 0.4) {
-						socket.emit('event', {
-							key: 'tower defense',
-							value: 'fire tower',
-							towerKey: tower.key,
-							unitKey: unit.key
-						});
-						break;
+	const fireTowers = useCallback((towerTypes: String[]) => {
+		towerTypes.forEach((type) => {
+			towerDefenseState.towers.filter((tower) => tower.type === type).forEach((tower) => {
```

Comment:
generally any time someone chains multiple loops together we should be cautious. This doesn't really affect our performance now but if there were a lot of types firing and towers placed, it could be a significant performance cost

here's one simpler workaround:

```
// towerTypes.forEach((type) => {
// 	towerDefenseState.towers.filter((tower) => tower.type === type).forEach((tower) => {
towerDefenseState.towers.forEach((tower) => {
				if (towerTypes.includes(tower.type)) {
```

If we wanted to be even more efficient, we would not use an array for the types because of O(n) search, instead we'd use an object (hash table)

like 

```
types = {
  'basic',
  'archer'
  }
  
  if(types[tower.type]) { // instead of .includes()
  ...
  }
  ```

Reply:
good call & thanks for the suggestions - went w/ the simple workaround